WordPress/Classes/Services/ThemeService.m

```diff
@@ -0,0 +1,270 @@
+#import "ThemeService.h"
+
+#import "Blog.h"
+#import "RemoteTheme.h"
+#import "Theme.h"
+#import "ThemeServiceRemote.h"
+#import "WPAccount.h"
+
+@implementation ThemeService
+
+#pragma mark - Themes availability
+
+- (BOOL)accountSupportsThemeServices:(WPAccount *)account
+{
+    NSParameterAssert([account isKindOfClass:[WPAccount class]]);
+    
+    return [account isWPComAccount];
+}
+
+- (BOOL)blogSupportsThemeServices:(Blog *)blog
+{
+    NSParameterAssert([blog isKindOfClass:[Blog class]]);
+    
+    return blog.restApi && [blog dotComID];
+}
+
+#pragma mark - Local queries: Creating themes
+
+/**
+ *  @brief      Creates and initializes a new theme with the specified theme Id in the specified
+ *              context.
+ *  @details    You should probably not call this method directly.  Please read the documentation
+ *              for findOrCreateThemeWithId: first.
+ *
+ *  @param      themeId     The ID of the new theme.  Cannot be nil.
+ *
+ *  @returns    The newly created and initialized object.
+ */
+- (Theme *)newThemeWithId:(NSString *)themeId
+{
+    NSParameterAssert([themeId isKindOfClass:[NSString class]]);
+    
+    NSEntityDescription *entityDescription = [NSEntityDescription entityForName:[Theme entityName]
+                                                         inManagedObjectContext:self.managedObjectContext];
+    
+    Theme *theme = [[Theme alloc] initWithEntity:entityDescription
+                  insertIntoManagedObjectContext:self.managedObjectContext];
+    
+    return theme;
+}
+
+/**
+ *  @brief      Obtains the theme with the specified ID if it exists, otherwise a new theme is
+ *              created and returned.
+ *
+ *  @param      themeId     The ID of the theme to retrieve.  Cannot be nil.
+ *
+ *  @returns    The stored theme matching the specified ID if found, or nil if it's not found.
+ */
+- (Theme *)findOrCreateThemeWithId:(NSString *)themeId
+{
+    NSParameterAssert([themeId isKindOfClass:[NSString class]]);
+    
+    Theme *theme = [self findThemeWithId:themeId];
+    
+    if (!theme) {
+        theme = [self newThemeWithId:themeId];
+    }
+    
+    return theme;
+}
+
+#pragma mark - Local queries: finding themes
+
+- (Theme *)findThemeWithId:(NSString *)themeId
+{
+    NSParameterAssert([themeId isKindOfClass:[NSString class]]);
+    
+    Theme *theme = nil;
+    
+    NSPredicate *predicate = [NSPredicate predicateWithFormat:@""];
+    NSFetchRequest *fetchRequest = [NSFetchRequest fetchRequestWithEntityName:[Theme entityName]];
+    
+    fetchRequest.predicate = predicate;
+    
+    NSError *error = nil;
+    NSArray *results = [self.managedObjectContext executeFetchRequest:fetchRequest error:&error];
+    
+    if (results) {
+        theme = (Theme *)[results firstObject];
+        NSAssert([theme isKindOfClass:[Theme class]],
```

Comment:
Beware: `executeFetchRequest:` may return an empty array (that is, results != nil, yet, count == 0) if there are no matching entities.

[Reference here](https://developer.apple.com/library/prerelease/ios/documentation/Cocoa/Reference/CoreDataFramework/Classes/NSManagedObjectContext_Class/index.html#//apple_ref/occ/instm/NSManagedObjectContext/executeFetchRequest:error:)